File: 11_Dropout/main.cpp
#include <iostream>
#include <opencv2/core/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <ctime>

using namespace cv;

Mat dropout(Mat& I, double p){
	Mat mask(I.size(),cv::DataType<float>::type);
	cv::randu(mask,0.0,1.0);
	cv::threshold(mask,mask,p,1.0,cv::THRESH_BINARY);
	return I.mul(mask);
}

int main(){
	cv::theRNG().state = time(0);
	Mat I(5,5,cv::DataType<float>::type);
	cv::randu(I,0,1.0);

	std::cout << I << std::endl;
	std::cout << dropout(I,0.5) << std::endl;

}

File: 11_Dropout/main.py
import cv2
import numpy as np

def dropout(i,p):
    mask = np.empty(i.shape, dtype=np.int16)
    mask = cv2.randu(mask,0,255)

    val, mask = cv2.threshold(mask,p*255,255,cv2.THRESH_BINARY)
    mask = np.asarray(mask,dtype=np.float64) / 255.0
    return cv2.multiply(i,mask)

I = np.random.rand(5,5)
O = dropout(I,0.5)

print I

print O
